File: leetcode/200.number-of-islands.py
# @leet start
class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:
        LAND = "1"

        rows = len(grid)
        cols = len(grid[0])

        count = 0

        def dfs(x: int, y: int) -> None:
            """Set all connected land to '-1'."""
            stack = [(x, y)]
            grid[x][y] = "-1"

            while stack:
                row, col = stack.pop()

                for delta_row, delta_col in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                    adj_row = row + delta_row
                    adj_col = col + delta_col

                    if (
                        0 <= adj_row < rows
                        and 0 <= adj_col < cols
                        and grid[adj_row][adj_col] == LAND
                    ):
                        stack.append((adj_row, adj_col))
                        grid[adj_row][adj_col] = "-1"

        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == LAND:
                    dfs(row, col)
                    count += 1

        return count


# @leet end

File: leetcode/200.number-of-islands.js
// @leet start
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  const DIRS = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];
  const LAND = "1";
  const EXPLORED = "-1";
  const m = grid.length;
  const n = grid[0].length;

  // Explore all adjacent land and mark processed land as water
  const dfs = (startR, startC) => {
    grid[startR][startC] = EXPLORED;
    const stack = [[startR, startC]]; // [number, number][]

    while (stack.length) {
      const [r, c] = stack.pop();

      for (const [dr, dc] of DIRS) {
        const nextR = r + dr;
        const nextC = c + dc;

        if (
          nextR >= 0 &&
          nextR < m &&
          nextC >= 0 &&
          nextC < n &&
          grid[nextR][nextC] === LAND
        ) {
          grid[nextR][nextC] = EXPLORED;
          stack.push([nextR, nextC]);
        }
      }
    }
  };

  let res = 0;

  for (let r = 0; r < m; r++) {
    for (let c = 0; c < n; c++) {
      if (grid[r][c] === LAND) {
        res++;
        dfs(r, c);
      }
    }
  }

  return res;
};
// @leet end
